### PR TITLE
chore(build): compile lua only on release builds

### PR DESCRIPTION
**Pull request by shadmansaleh**
Compile lua files only on release builds. Compiling them removes line number from tracebacks and makes them harder to debug

### CODE REVIEW

**Comment by zeertzjq** [comment]
I guess this will break tests?

File: CMakeLists.txt
Comment (gpanders) [inline CMakeLists.txt:548]: Should probably make this "not equal Debug" rather than "equal Release"

**Comment by gpanders** [comment]
Yes this will likely break some tests, although #16631 changed a lot of tests to not depend on the exact output of error traceback. So it shouldn't be too bad.

File: CMakeLists.txt
Comment (shadmansaleh) [reply CMakeLists.txt:548]: Which side does RelWithDebug go?

**Comment by zeertzjq** [comment]
I think the `any` key of `Screen:expect` can be used.

File: CMakeLists.txt
Comment (gpanders) [reply CMakeLists.txt:548]: Hmmm. I think since it’s a release build it should still be enabled. It can always be disabled manually with the Cmake option

**Comment by zeertzjq** [comment]
I'm trying to make that failing test more stable in #16860.

File: CMakeLists.txt
Comment (shadmansaleh) [reply CMakeLists.txt:548]: Ok .

Comment (shadmansaleh) [reply CMakeLists.txt]: Done.

**Comment by shadmansaleh** [comment]
@zeertzjq I've checked #16860  indeed fixes the test issue here . Thanks . This can be merged after #16860 is merged.

File: CMakeLists.txt
Comment (gpanders) [inline CMakeLists.txt:548]: This makes it impossible for someone to compile byte code on a debug build even if they wanted to. Instead, we should set the default value of COMPILE_LUA based on build type.

Debug: COMPILE_LUA = OFF
else: COMPILE_LUA = ON

This way it’s still user-overridable.

**Review by gpanders — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (zeertzjq) [inline CMakeLists.txt:546]: `DEBUG` is set to `0` or `1` above at lines 126 to 130.
```suggestion
if(DEBUG)
  option(COMPILE_LUA "Pre-compile Lua sources into bytecode (for sources that are included in the binary)" OFF)
else()
  option(COMPILE_LUA "Pre-compile Lua sources into bytecode (for sources that are included in the binary)" ON)
endif()
```

Comment (shadmansaleh) [reply CMakeLists.txt]: Thanks I had actually tried something similar earlier today but it didn't work for some reason back then . This works fine .

File: CMakeLists.txt
Comment (gpanders) [inline CMakeLists.txt:550]: Is there a way to do this without repeating the option name and description?

Comment (zeertzjq) [reply CMakeLists.txt:550]: Another option `ENABLE_LTO` also has its name and description repeated, so I think no.

**Review by gpanders — APPROVED** [review]

**Pull request merged**
